apiserver/facades/client/resources/charmhubclient_test.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package resources
+
+import (
+	"github.com/golang/mock/gomock"
+	"github.com/juju/charm/v8"
+	charmresource "github.com/juju/charm/v8/resource"
+	jc "github.com/juju/testing/checkers"
+	"github.com/juju/utils/v2/hash"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/apiserver/facades/client/resources/mocks"
+	"github.com/juju/juju/charmhub/transport"
+	corecharm "github.com/juju/juju/core/charm"
+)
+
+var _ = gc.Suite(&CharmHubClientSuite{})
+
+type CharmHubClientSuite struct {
+	client *mocks.MockCharmHub
+}
+
+func (s *CharmHubClientSuite) TestResolveResources(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+	s.client = mocks.NewMockCharmHub(ctrl)
+	s.expectRefresh()
+
+	result, err := s.newClient().ResolveResources(nil)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(result, gc.DeepEquals, []charmresource.Resource{{
+		Meta:     charmresource.Meta{Name: "wal-e", Type: 1, Path: "", Description: ""},
+		Origin:   2,
+		Revision: 0,
+		Fingerprint: charmresource.Fingerprint{
+			Fingerprint: hash.Fingerprint{}},
+		Size: 0,
+	}})
+}
+
+func (s *CharmHubClientSuite) TestResolveResourcesUpload(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+	s.client = mocks.NewMockCharmHub(ctrl)
+	s.expectRefresh()
+
+	result, err := s.newClient().ResolveResources([]charmresource.Resource{{
+		Meta:     charmresource.Meta{Name: "wal-e", Type: 1, Path: "", Description: ""},
+		Origin:   1,
+		Revision: 3,
+		Fingerprint: charmresource.Fingerprint{
+			Fingerprint: hash.Fingerprint{}},
+		Size: 0,
+	}})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(result, gc.DeepEquals, []charmresource.Resource{{
+		Meta:     charmresource.Meta{Name: "wal-e", Type: 1, Path: "", Description: ""},
+		Origin:   1,
+		Revision: 3,
+		Fingerprint: charmresource.Fingerprint{
+			Fingerprint: hash.Fingerprint{}},
+		Size: 0,
+	}})
+}
+
+func (s *CharmHubClientSuite) newClient() NewCharmRepository {
+	curl, _ := charm.ParseURL("ubuntu")
```

Comment:
We can use MustParseURL and MustParseChannel.

Reply:
done